File: frontend/src/types/construct.ts
export enum ElementType {
  PROMOTER = "promoter",
  KOZAK = "kozak",
  CDS = "cds",
  TERMINATOR = "terminator",
  TAG = "tag",
  UTR = "utr",
  CUSTOM = "custom",
}

export interface ConstructElement {
  element_type: ElementType;
  label: string;
  sequence: string;
  position: number;
  metadata_json?: Record<string, unknown>;
}

export interface Construct {
  id: string;
  project_id: string;
  name: string;
  full_sequence?: string;
  organism_tax_id?: number;
  elements: ConstructElement[];
  created_at: string;
  updated_at: string;
}

export interface ConstructCreate {
  project_id: string;
  name: string;
  organism_tax_id?: number;
  elements: ConstructElement[];
}

File: backend/app/schemas/construct.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field

from app.models.construct_element import ElementType


class ConstructElementSchema(BaseModel):
    element_type: ElementType
    label: str = Field(..., max_length=255)
    sequence: str = Field(..., max_length=100_000)
    position: int
    metadata_json: dict | None = None


class ConstructCreate(BaseModel):
    project_id: uuid.UUID
    name: str = Field(..., max_length=255)
    organism_tax_id: int | None = None
    elements: list[ConstructElementSchema] = []


class ConstructUpdate(BaseModel):
    name: str | None = Field(None, max_length=255)
    organism_tax_id: int | None = None
    elements: list[ConstructElementSchema] | None = None


class ConstructElementLabelUpdate(BaseModel):
    label: str = Field(..., max_length=255)


class RestrictionDigestRequest(BaseModel):
    enzymes: list[str] = Field(..., min_length=1, max_length=20)


class ConstructResponse(BaseModel):
    id: uuid.UUID
    project_id: uuid.UUID
    name: str
    full_sequence: str | None
    organism_tax_id: int | None
    elements: list[ConstructElementSchema]
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}
